### PR TITLE
Native Structured Build Output – Fix native context menu

### PR DESCRIPTION
**Pull request by sevoku**
This includes some ContextMenu API changes

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components/ContextMenu.cs [hunk 1]
```diff
@@ -60,12 +60,12 @@ public void CollapseSeparators ()
 				Items.RemoveAt (Items.Count - 1);
 		}
 
-		public void Show (Gtk.Widget parent, Gdk.EventButton evt)
+		public void Show (Control parent, Gdk.EventButton evt)
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components/ContextMenu.cs]: This is public API change.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Control.cs [hunk 2]
```diff
@@ -91,7 +91,7 @@ public T GetNativeWidget<T> ()
 			if (nativeWidget is T)
 				return (T)nativeWidget;
 			else
-				throw new NotSupportedException ();
+				return (T)ConvertToType (typeof (T), nativeWidget);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Control.cs]: Every time this is called, it'll create a new embed.

Comment (sevoku) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Components/Control.cs]: Yes, that was bad. Reverted and fixed with 9201fdf.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components/ContextMenu.cs (hunk 1 above)
Comment (sevoku) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Components/ContextMenu.cs]: Yes it is, we need a better solution if we don't allow API changes for the next release.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components/ContextMenu.cs [hunk 3]
```diff
@@ -89,12 +89,32 @@ public void Show (Gtk.Widget parent, int x, int y, Action closeHandler, bool sel
 			ContextMenuExtensionsGtk.ShowContextMenu (parent, x, y, this, closeHandler, selectFirstItem);
 		}
 
+		public void Show (Xwt.Widget parent, int x, int y, Action closeHandler)
+		{
+			var gtkWidget = parent.Surface.NativeWidget as Gtk.Widget;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components/ContextMenu.cs]: I'd prefer `is` pattern matching like:
```
if (parent.Surface.NativeWidget is Gtk.Widget widget) {
    Show (gtkWidget, x, y, closeHandler);
    return;
}
#if MAC
if (parent.Surface.NativeWidget is AppKit.NSView view) {
    Show (view, x, y, closeHandler);
    return;
}
#endif
throw new NotSupportedException ();
```

Or using a switch with case patterns.

Comment (sevoku) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Components/ContextMenu.cs]: right, thanks!

**Comment by sevoku** [comment]
Actually this still breaks the API. It's not possible to pass a Control because of the ambiguous call, so it also breaks the unit tests: https://github.com/mono/monodevelop/blob/structured-build-output-native-xwt-context-menu-fix/main/src/addins/MonoDevelop.UnitTesting/Services/AbstractUnitTestEditorExtension.cs#L274

**Comment by sevoku** [comment]
API break is fixed now: https://github.com/mono/monodevelop/pull/3950/commits/9e8620c3587af9ffc2d1d6b2b5ead37fec2868b6

**Pull request merged**